### ISSUE TITLE
Running the BinaryNet example gives poor accuracy

### ISSUE THREAD

**Issue by sms821** [NeuromorphicProcessorProject/snn_toolbox#32]
Hi Bodo,
I tried running the BinaryNet example without making any changes to the config file. For some reason, the accuracy of the parsed model and of the converted SNN is very poor (5%). The accuracy of the converted SNN is even worse (5%). Also, I was expecting the weights in the 88.22_parsed.h5 file to be binarized as we have set "binarize_weights=True", but this is not the case. The weight values are full-precision floats. Am I missing something? Please guide me.
![screen shot 2019-03-04 at 12 04 27 am](https://user-images.githubusercontent.com/24283722/53711675-3b31db00-3e11-11e9-9bb4-b7c0b23a5726.png)

My lasagne and theano versions are 0.2.dev1 and 1.0.2 respectively and I have installed the snntoolbox from the github repo.
I am attaching a snapshot of the simulation for your reference.

**Comment by rbodo** [comment]
I've looked into the issue with low ANN accuracy after parsing, it's the same for me. Will not be able to fix this right now - hopefully later this week.

Regarding binary weights, this post should help:

https://github.com/NeuromorphicProcessorProject/snn_toolbox/issues/28#issuecomment-447341818

**Comment by sms821** [comment]
Thank you for your reply.
I looked at #28 and have another question. If batch normalization is making the weights continuous again, will it be okay to re-cast them as binary during inference? Otherwise, they won't retain their binary characteristic and there won't be much to gain by constraining the weights to discrete {-1, +1} values during training.

**Comment by rbodo** [comment]
It's difficult to give a definite answer here as these quantization methods are still actively researched. Of course you can try re-casting them as binary, which may very well work for MNIST, but probably not for CIFAR-10 (or not without significant drop in accuracy). You could also try to train the original model without batch-normalization (BN), but it will be tricky to get the network to converge like that.

It can be argued that "binary" networks that have BN layers in them are still efficient, especially when running on dedicated hardware, because even if the weight values are full precision again due to BN, they are at least discretized, which reduces the memory footprint. (Each layer has only as many different weight values as there are channels in the feature map.)

**Comment by YiZhangNUDT** [comment]
Hi @sms821,
I also meet a similar problem about the low accuracy of BinaryNet for CIFAR-10.
Do you find the method to fix it?
Thank you very much.
Yi
